### DISCUSSION TITLE
Unexpected q.args value for a ui.button()

### DISCUSSION THREAD

**Discussion by azim-b** [h2oai/wave#1160] [Q&A]
When I hit a button 'Create', let's say the name is `create`, `q.args['create']` is `True.`.

Then I perform some logic, and let's say the creation was successful.

Then, I display a message bar saying it was successful, and show a message that says 'redirecting in x seconds' where x goes from 10 to 0, and the `page` is saved in every iteration.

Finally, after the for loop completes, I "reload" the page, i.e. call `serve(q)`.

Primary problem:
- when `serve(q)` is called, and the function in subject is called again (as `@on('#admin')`), let's say the name is `admin()`, the value of `q.args['create']` is still `True.`. (**Expectation: False**) So, the *same* logic repeats again, like an infinite loop.
a. why is the `q.args` value of the button still true?
b. how do I prevent the infinite loop, i.e. **not** execute the create login again (since create is not hit this time)

Secondary question:
- is there a nicer way to "reload"? Other than calling `serve(q)`.

Example flow (with the fault):

https://user-images.githubusercontent.com/64787868/146716592-79c9b1bd-56b7-4fba-b520-9d26784ce117.mov

Code snippet that deals with this use case:

```py
# First message, indicating Success
widgets.extend([
  ui.inline([
    ui.text("Refreshing the page in 10 seconds. Or"),
    ui.button(label=' click here ', name='#admin', link=True),
    ui.text("to refresh now.")
   ])
])

# "Display" the message, i.e. save the Page
q.page['body_right'] = ui.form_card(box='body_right', items=[
*widgets,
])
await q.page.save()

# Repeat the same, in a loop, and update # of seconds remaining for auto reload
for sec in range(9, -1, -1):
    widgets.pop()
    widgets.extend([
        ui.inline([
            ui.text(f"Refreshing the page in {sec} seconds. Or"),
            ui.button(label=' click here ', name='#admin', link=True),
            ui.text("to refresh now.")
        ])
    ])
    time.sleep(1)
    q.page['body_right'] = ui.form_card(box='body_right', items=[
    *widgets,
    ])
    await q.page.save()
```

**Comment by Far0n** [comment]
> when serve(q) is called, and the function in subject is called again (as @on('#admin')), let's say the name is admin(), the value of q.args['create'] is still True.. (Expectation: False) So, the same logic repeats again, like an infinite loop.

That's expected behavior, because q.args still holds the information about the button click. So if you really want to call serve, you need to make sure that q.args is in the state you want, e.g. clearing q.args beforehand. That's why it might be better to avoid calling serve from app-code (it's suppossed to be the submit entry point).

> how do I prevent the infinite loop, i.e. not execute the create login again (since create is not hit this time)

One way is below. It's basically a state transition from nothing created -> something created & show countdown -> something created post countdown (either countdown completed or user hit "click here" => show main page)

**Side note**: avoid using *time.sleep*, because it blocks the running event loop

```python
import asyncio
from h2o_wave import Q, app, main, ui

async def create(q: Q):
    q.page['main'] = ui.form_card(box="1 1 5 5", items=[ui.progress(label="Creating Something")])
    await q.page.save()
    await q.sleep(2)

async def countdown(q: Q):
    for sec in range(9, -1, -1):
        widgets = []
        widgets.extend([
            ui.inline([
                ui.text(f"Refreshing the page in {sec} seconds. Or"),
                ui.button(label=' click here ', name='admin', link=True),
                ui.text("to refresh now.")
            ])
        ])
        await q.sleep(1)
        q.page['main'] = ui.form_card(box='1 1 5 5', items=[*widgets])
        await q.page.save()

async def main_page(q: Q):
    q.page['main'] = ui.form_card(box='1 1 5 5', items=[ui.text("Main Page")])
    await q.page.save()

@app("/")
async def serve(q: Q):
    q.page.drop()

    if q.args.admin:
        q.client.countdown.cancel()
        return await main_page(q)

    if not q.client.something_created:
        await create(q)
        q.client.something_created = True
        q.client.countdown = asyncio.create_task(countdown(q))
        try:
            await q.client.countdown
        except asyncio.CancelledError:
            print("User clicked 'click here'")
        return await main_page(q)

    await main_page(q)
```

https://user-images.githubusercontent.com/6949295/146747921-7bcdeb72-894e-4985-9235-ece822b71feb.mp4

**Comment by mturoci** [comment]
> Finally, after the for loop completes, I "reload" the page, i.e. call serve(q)

@azim-b what is the purpose of the "reload"? Running the scenario again? From my limited point of view, it seems like calling a regular python function would do the trick. This way you are free of hacks and infinite loops. Calling `serve` directly is not advised.

Note: The reason why you are getting `True` for button value is because you are simply calling `serve` with the same `q` object so it's no surprise.
